pkg/armhelpers/support-validator.go

```diff
@@ -0,0 +1,118 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT license.
+
+package armhelpers
+
+import (
+	"context"
+
+	"github.com/Azure/aks-engine/pkg/api"
+	"github.com/pkg/errors"
+	log "github.com/sirupsen/logrus"
+)
+
+type requiredImage struct {
+	image     api.AzureOSImageConfig
+	errorData error
+}
+
+//ValidateRequiredImages checks both master ang agent profiles and ensures these images are available in the environment
+func ValidateRequiredImages(ctx context.Context, location string, p *api.Properties, client AKSEngineClient) error {
+
+	missingVMImages := make(map[string]requiredImage)
+
+	imageInformation := getDistroVMImageInformation(p.MasterProfile.Distro)
+
+	if imageFetcher, ok := client.(VMImageFetcher); ok {
+		vmImage, err := imageFetcher.GetVirtualMachineImage(ctx, location, imageInformation.ImagePublisher, imageInformation.ImageOffer, imageInformation.ImageSku, imageInformation.ImageVersion)
+
+		if err != nil {
+			if p.MasterProfile.Distro == "" {
+				missingVMImages[string(api.Ubuntu)] = requiredImage{
+					image:     imageInformation,
+					errorData: err,
+				}
+
+			} else {
+				missingVMImages[string(p.MasterProfile.Distro)] = requiredImage{
+					image:     imageInformation,
+					errorData: err,
+				}
+			}
+		} else {
+			log.Infof("Image %s-%s-%s found in your environment", imageInformation.ImageOffer, imageInformation.ImageSku, *vmImage.Name)
+		}
+
+		for _, profile := range p.AgentPoolProfiles {
+
+			if profile.OSType == api.Windows {
+				imageInformaition := api.WindowsServer2019OSImageConfig
+
+				vmImage, err := imageFetcher.GetVirtualMachineImage(ctx, location, imageInformaition.ImagePublisher, imageInformaition.ImageOffer, imageInformaition.ImageSku, imageInformaition.ImageVersion)
+
+				if err != nil {
+					missingVMImages[string(api.Windows)] = requiredImage{
+						image:     imageInformaition,
+						errorData: err,
+					}
+				} else {
+					log.Infof("Image %s-%s-%s found in your environment", imageInformation.ImageOffer, imageInformation.ImageSku, *vmImage.Name)
+				}
+			} else {
+				imageInformaition := getDistroVMImageInformation(profile.Distro)
```

Comment:
Please refactor to below. Also, please consider removing all non-essential logging statements.

``` go
imageInformaition := getDistroVMImageInformation(profile.Distro)
if _, err := imageFetcher.GetVirtualMachineImage(ctx, location, imageInformaition.ImagePublisher, imageInformaition.ImageOffer, imageInformaition.ImageSku, imageInformaition.ImageVersion)
  var distro = string(api.Ubuntu)
  if p.MasterProfie.Distro != "" {
    distro = profile.Distro
  }
  missingVMImages[distro] = requiredImage{
    image:     imageInformaition,
    errorData: err,
}
```

Reply:
Better yet, change the function to do the following:

```go
for _, profile := range p.AgentPoolProfiles {
  var imgInfo api.AzureOSImageConfig
  var distro string
  if profile.OSType == api.Windows {
    imgInfo := api.WindowsServer2019OSImageConfig
    distro = string(api.Windows)
  } else {
    distro = string(api.Ubuntu)
    if profile.Distro != "" {
      distro = string(profile.Distro)
    }

    imgInfo := getDistroVMImageInformation(distro)
  }

  if _, err := imageFetcher.GetVirtualMachineImage(ctx, location, imgInfo.ImagePublisher, imgInfo.ImageOffer, imgInfo.ImageSku, imgInfo.ImageVersion)
    missingVMImages[distro] = requiredImage{
      image:     imageInformaition,
      errorData: err,
  }
}
```